Checkout: Allow non-admins to renew their subscriptions

drewblaisdell opened the pull request:
Fixes #3141 and #1870. Requires D1331-code.

Currently, only admins can purchase upgrades for a site. When an admin is demoted to e.g. a contributor, their purchases still show up on `/purchases`, but they cannot renew them.

This PR updates checkout so that non-admins can renew purchases.

**Testing**
- Create a site through `/start`. We'll refer to the user that created this site as **Alice**.
- Invite another user you have access to to be an administrator on this site through `/people/new/:site`. We'll call the user you invited **Bob**.
- Log in to **Bob** in an incognito window.
- **Bob**'s email account will have received an email for this invitation with a link to accept. Paste this link into the incognito window to accept the invite with **Bob**.
- Visit `/plans/:site` as **Bob** and purchase a plan.
- Using our internal admin tools, change the expiry of the subscription for this plan to very soon (< 2 weeks in the future).
- As **Alice**, visit `/people/team/:site`, click on **Bob**, and change the account's role to either contributor, author, or editor.
- In the incognito window for **Bob**, execute `localStorage.clear()` in the console and refresh the page.
- Visit `/purchases`.
- Click on the expiring plan.
- Assert that, as **Bob**, you can renew the plan.
- As **Alice**, visit `/people/team/:site`, click on **Bob**, and remove the account from the site.
- As **Bob**, visit `/purchases` and click on the plan.
- Assert that the component loads and information about the purchase is displayed, even though **Bob** is not a member of the site to which the purchase is attached.

- [x] Code review
- [x] Product review

stephanethomas commented:
Code looks good.

stephanethomas commented:
I also tested it and it works fine. Good job!

fabianapsimoes commented:
The scenario where a user is demoted seems to work well. When a user is removed from a site though seems to be broken though. In that case, I do see the subscription in the list in `/purchases`, but I can't see the details in Manage Purchases (`/purchases/:site/:subId`). The placeholders are never replaced by actual content, and there are no errors in the console.

drewblaisdell commented:
Thanks for looking, @stephanethomas and @fabianapsimoes, and thanks for the commit, @scruffian.

I added a commit to render `ManagePurchase` when the user is not a member of the site to which the purchase is attached. I took the message from @scruffian's commit to add messaging for this case, and updated the component to remove links to pages that the user wouldn't have access to.

*With auto-renew disabled*
![screen shot 2016-03-16 at 2 05 34 pm](https://cloud.githubusercontent.com/assets/1130674/13829100/56ebc4da-eb82-11e5-8671-d0b98962ec33.png)

*With auto-renew enabled*
![screen shot 2016-03-16 at 2 14 41 pm](https://cloud.githubusercontent.com/assets/1130674/13829120/802f5c62-eb82-11e5-9d5b-e482959ff8fa.png)

This PR now also fixes #1870.

scruffian commented:
I updated 9de324e with a couple of changes:
- The message should be about "renewals" not "cancellations"
- The message should appear when auto renew is disabled, or when the purchase has expired:

<img width="756" alt="screen shot 2016-03-17 at 13 30 41" src="https://cloud.githubusercontent.com/assets/275961/13847322/9071711a-ec44-11e5-8d99-02d745c6aee4.png">
<img width="669" alt="screen shot 2016-03-17 at 13 26 15" src="https://cloud.githubusercontent.com/assets/275961/13847321/90506a88-ec44-11e5-9b6c-d2951ed781fb.png">

fabianapsimoes commented:
Is there any reason why we're opening the link to contact support in another tab?

Do we know what happens with subscribed purchases? Are we able to renew them correctly in the case the user has been removed from the site? Asking because right now we're only showing the support message to unsubscribed purchases. If automatic renewals are doomed to fail, we might benefit from showing the link right away.

Tug commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -145,7 +184,7 @@ const ManagePurchase = React.createClass( {
 			{ id, payment: { creditCard } } = purchase;
 
-		if ( isExpired( purchase ) || isOneTimePurchase( purchase ) || isIncludedWithPlan( purchase ) ) {
+		if ( isExpired( purchase ) || isOneTimePurchase( purchase ) || isIncludedWithPlan( purchase ) || ! getSelectedSite( this.props ) ) {
```
Would be nice to have comment here or move this in a new function.
Same for other conditions in this file with 4 calls.

Tug replied:
Also I fail to see how `getSelectedSite( this.props )` is better than `this.props.selectedSite`. Sure it avoids some duplication of code but it hides the real name of the prop

Tug commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -283,6 +322,10 @@ const ManagePurchase = React.createClass( {
 			purchase = getPurchase( this.props );
 		let url, text;
 
+		if ( ! getSelectedSite( this.props ) ) {
```
Plus we already have `selectedSite` in this context here

scruffian commented:
I updated fbae85b to show this message in all cases then the user isn't on the site anymore:
<img width="536" alt="screen shot 2016-03-17 at 15 47 46" src="https://cloud.githubusercontent.com/assets/275961/13851873/bb343104-ec57-11e5-9972-314aaadb9161.png">

scruffian commented:
> Is there any reason why we're opening the link to contact support in another tab?

This is a pattern elsewhere in Calypso, but we can change it...

Tug commented:
Changes look good to me. I had a few minor comments because I find some part are becoming difficult to understand with all those ifs.

scruffian commented:
I added a021a65 to address some of @Tug's concerns. I am not sure this is an improvement.

drewblaisdell commented:
Thanks for the reviews, @fabianapsimoes / @Tug / @scruffian.

Per discussion elsewhere, we're not going to remove `getSelectedSite` in this PR because it would require testing `CancelPurchase` which isn't related to the two issues this fixes.

> Is there any reason why we're opening the link to contact support in another tab?

I updated this to open in the same window.

> Do we know what happens with subscribed purchases? Are we able to renew them correctly in the case the user has been removed from the site?

I'm fairly sure the answer is yes, but I'll investigate and update this PR when I find an answer.

fabianapsimoes commented:
> I'm fairly sure the answer is yes, but I'll investigate and update this PR when I find an answer.

Ben updated the PR to always show the message, here https://github.com/Automattic/wp-calypso/pull/4067#issuecomment-197942357.

drewblaisdell commented:
@fabianapsimoes I saw the change, but it feels a little strange to me to indicate that users should contact support to renew when the 'Auto-Renews On' label indicates that the purchase will be renewed anyway. What do you think about hiding the message if the purchase will renew successfully?

fabianapsimoes commented:
> What do you think about hiding the message if the purchase will renew successfully?

I definitely see what you mean. The idea was to try and prompt users to contact support mainly so they can sort out the ownership issues (since they've been removed / demoted from a site), not so much for the renewal itself. (See p1458228730000157-amber-slack)

drewblaisdell commented:
@fabianapsimoes Ah, okay, I agree with that rationale, and agree that the message makes sense (especially because the user can only renew early through support). :+1:

scruffian commented:
Product :+1:
